Added AD templating support for coupled neighbor variables

hirsja opened the pull request:
closes #23323

@lindsayad

hirsja committed:
Added AD templating support for coupled neighbor variables #23323

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1329086/) on ebab796 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23324/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1329100/) on ebab796 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c677e7</th>
      <th colspan=4>#23324 ebab79</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.62%</td>
      <td>84.62%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>81993</td>
      <td>81992</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14898</td>
      <td>14899</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23324/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23324/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23324/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented:
@hirsja I will try to look at this soon!

lindsayad committed:
Rename generic neighbor coupleable routine

The new name makes this more consistent with the non-neighbor
routine name: `coupledGenericValue`, although it is perhaps a little
less consistent with the analog AD routine `adCoupledNeighborValue`.

Also move the definition outside of the class definition

lindsayad committed:
Test new generic neighbor coupleable routine

lindsayad commented:
@hirsja hopefully you are ok with the routine rename? If you really don't like the rename, I'm open to changing it back.

Since I added the testing commit, I'm requesting independent review from @loganharbour or @cticenhour

hirsja commented:
@lindsayad Looks good to me, thanks for the review and revisions.
